### PR TITLE
Remove build directories from RPATH on FreeBSD

### PR DESCRIPTION
**Pull request by ararslan**
Currently we're putting build directories as well as GCC's directory directly into our `RPATH`. This is fine for simple source builds that sit where they are and aren't distributed, like on CI. However, this causes major issues for distributing.

~~Currently a work in progress, as the script works on its own but doesn't seem to be getting called properly during the build. Opening as a PR anyway to solicit suggestions for fixing that.~~ Fixed, thanks to Elliot.

### CODE REVIEW

**Review by staticfloat — APPROVED** [review]
Some small comments, overall looks good!

File: contrib/fixup-freebsd-rpath.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    rpath="$(${patchelf} --print-rpath ${lib})"
+
+    # If it doesn't contain the build's libdir, we're don't care about it
+    if [ -z "$(echo ${rpath} | grep -F ${build_libdir})" ]; then
+        continue
+    fi
+
+    # Remove build_libdir from the RPATH, retaining the rest
+    new_rpath="$(echo ${rpath} | tr : \\n | grep -vF ${build_libdir} | tr \\n :)"
+    new_rpath="${new_rpath%?}"
```
[the remark this answers is not in the document]
Comment (ararslan) [reply contrib/fixup-freebsd-rpath.sh]: Yep

File: deps/blas.mk [hunk 2]
```diff
@@ -64,6 +64,7 @@ endif
 
 OPENBLAS_BUILD_OPTS += CFLAGS="$(CFLAGS) $(OPENBLAS_CFLAGS)"
 OPENBLAS_BUILD_OPTS += FFLAGS="$(FFLAGS) $(OPENBLAS_FFLAGS)"
+OPENBLAS_BUILD_OPTS += LDFLAGS="$(LDFLAGS) $(RPATH_ESCAPED_ORIGIN)"
```
[the remark this answers is not in the document]
Comment (ararslan) [reply deps/blas.mk]: CI seemed okay with it. What other kind of testing would be appropriate here to ensure it's legit?

**Comment by nalimilan** [comment]
Why only on FreeBSD?

**Comment by ararslan** [comment]
That's the only system where this is an issue. We specifically have to add the build directories to the RPATH during the build on FreeBSD or else we end up linking to the system's outdated, incompatible libgcc_s. But we don't want the build directories in the RPATH for distribution, so we have to remove them at install time.

**Comment by ararslan** [comment]
AppVeyor failure is a timeout.

**Comment by iblis17** [comment]
I'm not a fan of copying GCC's binary in a simple source build.
(I'm curious about that we copy the GCC's lib and load against it, if there are some programs already loaded GCC's lib from system like `/usr/local/lib/...`, now we open julia, will the GCC libs being loaded into memory again?)
Is there a way to make the copying for distribution related `make` target only? (like `install` target only

**Comment by staticfloat** [comment]
> if there are some programs already loaded GCC's lib from system like `/usr/local/lib/...`, now we open julia, will the GCC libs being loaded into memory again?

`julia` is the process; so it is the first thing being loaded.  Every other dynamic library is being loaded into `julia`'s address space, so they all need to be using the same GCC libraries, which all must be the same as what Julia loads.  Of course, on most systems, once a library (such as `libgcc_s`) is loaded into an address space, if another library depends on that library (even if it contains an absolute path to a particular version) the dynamic linker won't even look for it.  That is, as long as the `SONAME` is the same.

> Is there a way to make the copying for distribution related `make` target only? (like `install` target only

We actually need this for the source build as well; on FreeBSD, for whatever reason `gfortran` does not embed the correct paths to its own runtime libraries within binaries it builds.  This means that if you compile a program, then immediately try to run it, it will fail, because it does not find the correct runtime libraries.  In the case of FreeBSD, it picks up the wrong `libgcc_s`.  Our solution is to copy the libraries that it should be loading into `$(build_privatelibdir)`.  I don't see any problems with this, as it is just copying in the files that we're trying to load in the first place, and also makes our `make binary-dist` steps much simpler.

File: deps/blas.mk (hunk 2 above)
[the remark this answers is not in the document]
Comment (staticfloat) [reply deps/blas.mk]: CI is pretty good, as long as it's separated testing.  I'm mostly scared of things working in the source directory and then breaking when we move it away.

[the remark this answers is not in the document]
Comment (ararslan) [reply deps/blas.mk]: I think most of our CI providers do testing outside of the source tree. At least Travis does.

**Pull request merged**